diff --git a/src/condor_scripts/adstash/ad_sources/schedd_transfer_epoch_history.py b/src/condor_scripts/adstash/ad_sources/schedd_transfer_epoch_history.py
index b0094d7d9e..dd37bb3429 100644
--- a/src/condor_scripts/adstash/ad_sources/schedd_transfer_epoch_history.py
+++ b/src/condor_scripts/adstash/ad_sources/schedd_transfer_epoch_history.py
@@ -39,8 +39,10 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
         self.indexed_keyword_attrs = {
             "Endpoint",
             "ErrorType",
+            "ErrorMessage",
             "FailedName",
             "FailureType",
+            "DebugErrorType",
             "GLIDEIN_ResourceName",
             "GLIDEIN_Site",
             "GlobalJobId",
@@ -75,6 +77,7 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
             "ErrorCode",
             "LibcurlReturnCode",
             "NumShadowStarts",
+            "PelicanErrorCode",
             "ProcId",
             "TransferFileBytes",
             "TransferHttpStatusCode",
@@ -90,6 +93,7 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
         }
         self.bool_attrs = {
             "FinalAttempt",
+            "Retryable",
             "TransferSuccess",
         }
         self.nested_attrs = set()
@@ -189,11 +193,14 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
 
     def expand_plugin_result_ads(self, ads, my_attr_name=None):
         debug_results = []
+        error_results = iter(())
         result = {}
         for ad in ads:
             for attr, value in ad.items():
                 if attr.lower() == "developerdata":
                     debug_results = self.expand_debug_ad(value)
+                elif attr.lower() == "transfererrordata":
+                    error_results = self.expand_error_ads(value)
                 else:
                     attr, value = self.normalize(attr, value)
                     result[attr] = value
@@ -205,18 +212,26 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
                 result["Attempts"] = 1
                 result["Attempt"] = 0
                 result["FinalAttempt"] = True
+                try:  # add error data if it exists
+                    result.update(next(error_results))
+                except StopIteration:
+                    pass
                 yield result
             else:
                 for debug_result in debug_results:
+                    try:
+                        error_result = next(error_results)
+                    except StopIteration:
+                        error_result = {}
                     if debug_result.get("FinalAttempt"):  # merge and return entire ad on final attempt
                         try:
-                            yield debug_result | result
+                            yield debug_result | result | error_result
                         except TypeError:  # backwards compat
                             yield {**debug_result, **result}
                     else:  # otherwise only add identifying attrs
                         for attr in ("TransferProtocol", "TransferType", "TransferUrl"):
                             debug_result[attr] = result.get(attr)
-                        yield debug_result
+                        yield debug_result | error_result
 
 
     def expand_debug_ad(self, ad):
@@ -276,6 +291,22 @@ class ScheddTransferEpochHistorySource(GenericAdSource):
                 yield {**result, **attempt_result}
 
 
+    def expand_error_ads(self, ads):
+        for ad in ads:
+            error_result = {}
+            for attr, value in ad.items():
+                if attr.lower() == "developerdata":
+                    for debug_attr, debug_value in value.items():
+                        if debug_attr in ad:
+                            debug_attr = f"Debug{debug_attr}"
+                        debug_attr, debug_value = self.normalize(debug_attr, debug_value)
+                        error_result[debug_attr] = debug_value
+                else:
+                    attr, value = self.normalize(attr, value)
+                    error_result[attr] = value
+            yield error_result
+
+
     def to_json_list(self, ad, return_dict=False):
         result = {}
 
diff --git a/src/condor_scripts/adstash/convert.py b/src/condor_scripts/adstash/convert.py
index b1922a08ea..31f4fe4b9d 100644
--- a/src/condor_scripts/adstash/convert.py
+++ b/src/condor_scripts/adstash/convert.py
@@ -81,7 +81,12 @@ INDEXED_KEYWORD_ATTRS = {
     "MATCH_EXP_JOB_GLIDEIN_ClusterId",
     "MATCH_EXP_JOB_GLIDEIN_Entry_Name",
     "MATCH_EXP_JOB_GLIDEIN_Factory",
+    "MATCH_EXP_JOB_GLIDEIN_Job_Max_Time",
+    "MATCH_EXP_JOB_GLIDEIN_Max_Walltime",
+    "MATCH_EXP_JOB_GLIDEIN_MaxMemMBs",
+    "MATCH_EXP_JOB_GLIDEIN_Memory",
     "MATCH_EXP_JOB_GLIDEIN_Name",
+    "MATCH_EXP_JOB_GLIDEIN_ProcId",
     "MATCH_EXP_JOB_GLIDEIN_Schedd",
     "MATCH_EXP_JOB_GLIDEIN_SEs",
     "MATCH_EXP_JOB_GLIDEIN_Site",
@@ -89,6 +94,8 @@ INDEXED_KEYWORD_ATTRS = {
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS_Queue",
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS_Slot",
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS",
+    "MATCH_EXP_JOB_GLIDEIN_ToDie",
+    "MATCH_EXP_JOB_GLIDEIN_ToRetire",
     "MATCH_EXP_JOBGLIDEIN_ResourceName",
     "MyType",
     "NTDomain",
@@ -249,13 +256,6 @@ INT_ATTRS = {
     "MachineAttrCpus0",
     "MachineAttrSlotWeight0",
     "MachineCount",
-    "MATCH_EXP_JOB_GLIDEIN_Job_Max_Time",
-    "MATCH_EXP_JOB_GLIDEIN_Max_Walltime",
-    "MATCH_EXP_JOB_GLIDEIN_MaxMemMBs",
-    "MATCH_EXP_JOB_GLIDEIN_Memory",
-    "MATCH_EXP_JOB_GLIDEIN_ProcId",
-    "MATCH_EXP_JOB_GLIDEIN_ToDie",
-    "MATCH_EXP_JOB_GLIDEIN_ToRetire",
     "MaxHosts",
     "MaxJobRetirementTime",
     "MaxTransferInputMB",
diff --git a/src/condor_scripts/adstash/interfaces/elasticsearch.py b/src/condor_scripts/adstash/interfaces/elasticsearch.py
index 988dc0c64f..5ea68caa3c 100644
--- a/src/condor_scripts/adstash/interfaces/elasticsearch.py
+++ b/src/condor_scripts/adstash/interfaces/elasticsearch.py
@@ -134,7 +134,7 @@ class ElasticsearchInterface(GenericInterface):
                 "adstash_runtime": {"type": "date", "format": "epoch_second"},
                 "condor_history_runtime": {"type": "date", "format": "epoch_second"},
             },
-            "type": "nested",
+            "type": "object",
             "dynamic": True,
         }
 
@@ -155,7 +155,7 @@ class ElasticsearchInterface(GenericInterface):
             for name in convert.BOOL_ATTRS:
                 properties[name] = {"type": "boolean"}
             for name in convert.NESTED_ATTRS:
-                properties[name] = {"type": "nested", "dynamic": True}
+                properties[name] = {"type": "object", "dynamic": True}
 
             dynamic_templates = [
                 {
@@ -190,6 +190,15 @@ class ElasticsearchInterface(GenericInterface):
                         "mapping": {"type": "boolean"},
                     }
                 },
+                {
+                    "transfer_stats" : {
+                        "match" : "^Transfer(Input|Output|Checkpoint)Stats\\.*$",
+                        "match_pattern" : "regex",
+                        "mapping" : {
+                            "type" : "long"
+                        }
+                    }
+                },
                 {
                     "strings_as_keywords": {  # Store unknown strings as keywords
                         "match_mapping_type": "string",
@@ -214,7 +223,7 @@ class ElasticsearchInterface(GenericInterface):
             for name in ad_source.bool_attrs:
                 properties[name] = {"type": "boolean"}
             for name in ad_source.nested_attrs:
-                properties[name] = {"type": "nested", "dynamic": True}
+                properties[name] = {"type": "object", "dynamic": True}
             dynamic_templates = ad_source.dynamic_templates
 
         mappings = {
@@ -284,6 +293,8 @@ class ElasticsearchInterface(GenericInterface):
                 mappings_file = log_dir / "condor_adstash_elasticsearch_last_mappings.json"
                 logging.debug(f"Writing updated mappings to {mappings_file}.")
                 json.dump(updated_mappings, open(mappings_file, "w"), indent=2)
+                old_mappings_file = log_dir / "condor_adstash_elasticsearch_previous_mappings.json"
+                json.dump(existing_mappings, open(old_mappings_file, "w"), indent=2)
 
 
     def make_bulk_body(self, ads, metadata={}):
