diff --git a/src/condor_scripts/adstash/config.py b/src/condor_scripts/adstash/config.py
index 81492f4026..9e076552cb 100644
--- a/src/condor_scripts/adstash/config.py
+++ b/src/condor_scripts/adstash/config.py
@@ -71,6 +71,7 @@ def get_htcondor_config(name="ADSTASH"):
         "threads": p.get(f"{name}_NUM_THREADS"),
         "collectors": p.get(f"{name}_READ_POOLS"),
         "schedds": p.get(f"{name}_READ_SCHEDDS"),
+        "ignore_schedds": p.get(f"{name}_IGNORE_SCHEDDS"),
         "startds": p.get(f"{name}_READ_STARTDS"),
         "read_schedd_history": p.get(f"{name}_SCHEDD_HISTORY"),
         "read_startd_history": p.get(f"{name}_STARTD_HISTORY"),
@@ -140,6 +141,7 @@ def get_environment_config(name="ADSTASH"):
         "threads": env.get(f"{name}_NUM_THREADS"),
         "collectors": env.get(f"{name}_READ_POOLS"),
         "schedds": env.get(f"{name}_READ_SCHEDDS"),
+        "ignore_schedds": env.get(f"{name}_IGNORE_SCHEDDS"),
         "startds": env.get(f"{name}_READ_STARTDS"),
         "read_schedd_history": env.get(f"{name}_SCHEDD_HISTORY"),
         "read_startd_history": env.get(f"{name}_STARTD_HISTORY"),
@@ -439,6 +441,13 @@ def get_config(argv=None):
             "[default is to process all Schedds located by Collectors]"
         ),
     )
+    history_group.add_argument(
+        "--ignore-schedds",
+        help=(
+            "Comma-separated list of Schedd names to skip processing "
+            "[default is to process all Schedds located by Collectors]"
+        ),
+    )
     history_group.add_argument(
         "--startds",
         help=(
diff --git a/src/condor_scripts/adstash/convert.py b/src/condor_scripts/adstash/convert.py
index b1922a08ea..31f4fe4b9d 100644
--- a/src/condor_scripts/adstash/convert.py
+++ b/src/condor_scripts/adstash/convert.py
@@ -81,7 +81,12 @@ INDEXED_KEYWORD_ATTRS = {
     "MATCH_EXP_JOB_GLIDEIN_ClusterId",
     "MATCH_EXP_JOB_GLIDEIN_Entry_Name",
     "MATCH_EXP_JOB_GLIDEIN_Factory",
+    "MATCH_EXP_JOB_GLIDEIN_Job_Max_Time",
+    "MATCH_EXP_JOB_GLIDEIN_Max_Walltime",
+    "MATCH_EXP_JOB_GLIDEIN_MaxMemMBs",
+    "MATCH_EXP_JOB_GLIDEIN_Memory",
     "MATCH_EXP_JOB_GLIDEIN_Name",
+    "MATCH_EXP_JOB_GLIDEIN_ProcId",
     "MATCH_EXP_JOB_GLIDEIN_Schedd",
     "MATCH_EXP_JOB_GLIDEIN_SEs",
     "MATCH_EXP_JOB_GLIDEIN_Site",
@@ -89,6 +94,8 @@ INDEXED_KEYWORD_ATTRS = {
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS_Queue",
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS_Slot",
     "MATCH_EXP_JOB_GLIDEIN_SiteWMS",
+    "MATCH_EXP_JOB_GLIDEIN_ToDie",
+    "MATCH_EXP_JOB_GLIDEIN_ToRetire",
     "MATCH_EXP_JOBGLIDEIN_ResourceName",
     "MyType",
     "NTDomain",
@@ -249,13 +256,6 @@ INT_ATTRS = {
     "MachineAttrCpus0",
     "MachineAttrSlotWeight0",
     "MachineCount",
-    "MATCH_EXP_JOB_GLIDEIN_Job_Max_Time",
-    "MATCH_EXP_JOB_GLIDEIN_Max_Walltime",
-    "MATCH_EXP_JOB_GLIDEIN_MaxMemMBs",
-    "MATCH_EXP_JOB_GLIDEIN_Memory",
-    "MATCH_EXP_JOB_GLIDEIN_ProcId",
-    "MATCH_EXP_JOB_GLIDEIN_ToDie",
-    "MATCH_EXP_JOB_GLIDEIN_ToRetire",
     "MaxHosts",
     "MaxJobRetirementTime",
     "MaxTransferInputMB",
diff --git a/src/condor_scripts/adstash/interfaces/elasticsearch.py b/src/condor_scripts/adstash/interfaces/elasticsearch.py
index 988dc0c64f..5ea68caa3c 100644
--- a/src/condor_scripts/adstash/interfaces/elasticsearch.py
+++ b/src/condor_scripts/adstash/interfaces/elasticsearch.py
@@ -134,7 +134,7 @@ class ElasticsearchInterface(GenericInterface):
                 "adstash_runtime": {"type": "date", "format": "epoch_second"},
                 "condor_history_runtime": {"type": "date", "format": "epoch_second"},
             },
-            "type": "nested",
+            "type": "object",
             "dynamic": True,
         }
 
@@ -155,7 +155,7 @@ class ElasticsearchInterface(GenericInterface):
             for name in convert.BOOL_ATTRS:
                 properties[name] = {"type": "boolean"}
             for name in convert.NESTED_ATTRS:
-                properties[name] = {"type": "nested", "dynamic": True}
+                properties[name] = {"type": "object", "dynamic": True}
 
             dynamic_templates = [
                 {
@@ -190,6 +190,15 @@ class ElasticsearchInterface(GenericInterface):
                         "mapping": {"type": "boolean"},
                     }
                 },
+                {
+                    "transfer_stats" : {
+                        "match" : "^Transfer(Input|Output|Checkpoint)Stats\\.*$",
+                        "match_pattern" : "regex",
+                        "mapping" : {
+                            "type" : "long"
+                        }
+                    }
+                },
                 {
                     "strings_as_keywords": {  # Store unknown strings as keywords
                         "match_mapping_type": "string",
@@ -214,7 +223,7 @@ class ElasticsearchInterface(GenericInterface):
             for name in ad_source.bool_attrs:
                 properties[name] = {"type": "boolean"}
             for name in ad_source.nested_attrs:
-                properties[name] = {"type": "nested", "dynamic": True}
+                properties[name] = {"type": "object", "dynamic": True}
             dynamic_templates = ad_source.dynamic_templates
 
         mappings = {
@@ -284,6 +293,8 @@ class ElasticsearchInterface(GenericInterface):
                 mappings_file = log_dir / "condor_adstash_elasticsearch_last_mappings.json"
                 logging.debug(f"Writing updated mappings to {mappings_file}.")
                 json.dump(updated_mappings, open(mappings_file, "w"), indent=2)
+                old_mappings_file = log_dir / "condor_adstash_elasticsearch_previous_mappings.json"
+                json.dump(existing_mappings, open(old_mappings_file, "w"), indent=2)
 
 
     def make_bulk_body(self, ads, metadata={}):
diff --git a/src/condor_scripts/adstash/utils.py b/src/condor_scripts/adstash/utils.py
index d81229808c..6ad295d471 100644
--- a/src/condor_scripts/adstash/utils.py
+++ b/src/condor_scripts/adstash/utils.py
@@ -53,6 +53,8 @@ def get_schedds(args):
         for schedd in schedds:
             if args.schedds and not (schedd["Name"] in args.schedds.split(",")):
                 continue
+            if args.ignore_schedds and (schedd["Name"] in args.ignore_schedds.split(",")):
+                continue
             schedd["MyPool"] = host
             try:
                 schedd_ads[schedd["Name"]] = schedd
